File: 2020/fsharp/Common.Test/UnitTests/Navigation.fs
namespace UnitTests

module Navigation =
    open Common.Navigation
    open FsUnit.Xunit
    open Xunit

    [<Fact>]
    let ``Navigate directions`` () =
        let directions = [North; North; East; West; West; South; North; West; West; West; East]
        List.fold move (0, 0) directions |> should equal (-3, 2)

    [<Fact>]
    let ``Navigate rotations`` () =
        let rotations = [(5, Right); (3, Left); (2, Left); (6, Left); (4, Right)]
        let takeSteps (position, direction) (steps, rotation) = 
            let newPosition = List.replicate steps direction |> List.fold move position
            let newDirection = turn direction rotation
            (newPosition, newDirection)

        List.fold takeSteps ((0, 0), North) rotations |> should equal ((-3, 3), West)

File: 2021/kotlin/src/main/kotlin/nl/sanderp/aoc/common/Navigation.kt
package nl.sanderp.aoc.common

enum class Direction { North, East, South, West }

private val cw = listOf(Direction.North, Direction.East, Direction.South, Direction.West)
private val ccw = listOf(Direction.North, Direction.West, Direction.South, Direction.East)

fun Direction.turnClockwise(times: Int = 1) = turn(cw, times)

fun Direction.turnCounterClockwise(times: Int = 1) = turn(ccw, times)

private fun Direction.turn(turns: List<Direction>, times: Int): Direction {
    val id = (turns.indexOf(this) + times) % turns.size
    return turns[id]
}

fun Point2D.move(direction: Direction, steps: Int = 1) = when (direction) {
    Direction.North -> x to y + steps
    Direction.East -> x + steps to y
    Direction.South -> x to y - steps
    Direction.West -> x - steps to y
}